enhancements/network/openstack-dual-stack.md

```diff
@@ -0,0 +1,188 @@
+
+## Motivation
+
+Customers need dual-stack support to allow migration of workloads from IPv4 to IPv6
+as they are running out of IPv4 addresses. This will allow OpenShift on OpenStack to
+serve a wider variety of workloads.
+
+### User Stories
+
+* As an user of a IPI OpenShift cluster, I want to have ingress and egress working with dual-stack.
+* As an user of a IPI OpenShift cluster, I want clusterIP services to be working with dual-stack.
+
+### Goals
+
+- OpenStack VMs are configured with one dual-stack interface.
+- The IPv6 Subnets can be configured with: Stateless, SLAAC or Stateful.
+- The addresses on the VM's interface are configured with the same address
+assigned by OpenStack to the port.
+- Nodes are configured with both IPv4 and IPv6 addresses, without exposing
+additional addresses, like storage interfaces.
+- Pods are configured with both IPv4 and IPv6 addresses.
+
+### Non-Goals
+
+- Support for OpenStack endpoints configured with dual-stack.
+- Support for LoadBalancer type Services.
+- Installer (IPI) handles creation of the Network and dual-stack Subnet.
+
+## Proposal
+
+Customers will be able to specify the IPv4 and IPv6 Subnets
+meant to be used in the `install-config.yaml`.
+
+### Workflow Description
+
+1. The customers will configure a network (tenant or provider) with IPv4 and
+IPv6 subnets. The IPv6 address mode for the Subnet can be of any type:
+SLAAC, Stateless or Stateful. Here is a sample of network and subnets creation:
+
+```bash
+$ openstack network create dual-stack
+$ openstack subnet create --network dual-stack --subnet-range 10.0.0.0/16 ipv4-subnet
+$ openstack subnet create --network dual-stack --ipv6-address-mode slaac --ipv6-ra-mode slaac --subnet-range 2001:db8:2222:5555::/64 --ip-version 6 ipv6-subnet
+```
+
+2. The customer will create API and Ingress Ports using the network created in the
+previous step. Here is a sample of ports creation:
+
+```bash
+$ openstack port create api --network dual-stack
+$ openstack port create ingress --network dual-stack
+```
+
+3. The customer will write the `install-config.yaml` and add the subnet IDs to
+the new field `machinesSubnets`.
+4. The customer will generate the manifests and configured MachineConfigs
+for the master and workers depending on the IPv6 addresses mode chosen for the Subnets
+in step 1. If SLAAC or Stateless was used the MachineConfigs has to contain `ip_addr_mode=eui64`
+and `method=auto` in the connection for the ethernet interface. While when using stateful
+it has to enable `ip=dhcp,dhcpv6` in the kernel arguments.
+5. The `uuid` from the network has to be added in the worker MachineSet alongside with the
+subnets that were pre-filled for the machine-api-provider-openstack to create one unique Port
+with both addresses configured for the Server.
+6. The CIDRs of the subnets have to added to the `cloud-provider-config.yaml`
+in the `address-sort-order` field to make sure IPv4 address will have higher priority in the
+Node addresses.
```

Comment:
That is:
* Why are we doing this by CIDR?
* Why do IPv4 addresses need to come first?

Reply:
@mdbooth openshift didn't allow IPv6 to be the primary Subnet, there was no support for it. Only recently I learned that IPv6 as the primary was added in 4.13 (I think) and I can now double check if it works for us.
I controlled this by CIDR in a CPO configuration because this would assure that the nodeAddresses are provided in the correct order.